contrib/ros/envs/sawyer_env.py

```diff
@@ -0,0 +1,356 @@
+                 sparse_reward,
+                 target_in_the_air,
+                 target_range=0.15,
+                 obj_range=0.15):
+        """
+        :param initial_robot_joint_pos: {str: float}
+                {'joint_name': value}
+        :param robot control mode: str
+                'effort'/'position'/'velocity'
+        :param real: bool
+                if the environment is for real robot or simulation
+        :param has_object: bool
+                if there is object in this experiment
+        :param distance_threshold: float
+                When the distance from robot endpoint to the target point is smaller than this threshold
+                the training episode is done.
+                To calculate the sparse reward, if the distance is smaller than this threshold returns -1
+                else returns 0.
+        :param initial_model_pos: dict
+                joint positions and object positions
+        :param sparse_reward: bool
+                if reward_type is 'sparse'
+        :param target_in_the_air: bool
+                set to see if the target can be in the air
+        :param target_range: float
+                the new target position for the new training episode would be randomly sampled inside
+                [initial_gripper_pos - target_range, initial_gripper_pos + target_range]
+        :param obj_range: float
+                the new initial object x,y position for the new training episode would be randomly sampled inside
+                [initial_gripper_pos - obj_range, initial_gripper_pos + obj_range]
+        """
+        # run the superclass initialize function first
+        Serializable.quick_init(self, locals())
+
+        logger.log('Initializing ros node!')
+        rospy.init_node('SawyerEnv', anonymous=True)
+        # Verify robot is enabled
+        self.limb = intera_interface.Limb('right')
+        logger.log('Initializing robot!')
+        self._robot = Sawyer(
+            initial_robot_joint_pos, control_mode=robot_control_mode)
+        if not self._robot.enabled:
+            logger.log("Please enable the sawyer!")
+            raise Exception
+        else:
+            logger.log("Robot has been enabled!")
+
+        self._real = real
+        if not self._real:
+            rospy.Subscriber('/gazebo/model_states', ModelStates,
+                             self.update_model_states)
+            self._gazebosrv = Gazebo()
+        else:
+            # rospy.Subscriber('/gazebo/model_states', ModelStates,
+            #                  self.update_model_states)
+            # TODO
+            pass
+        self._initial_model_pos = initial_model_pos
+        self.has_object = has_object
+        self.distance_threshold = distance_threshold
+        self.sparse_reward = sparse_reward
+        self.target_range = target_range
+        self.target_in_the_air = target_in_the_air
+        self.obj_range = obj_range
+
+        def clean_shutdown():
```

Comment:
Please define this as a private function of the class, rather than as a nested function. Call it `shutdown()`.

Reply:
`shutdown()` might also be part of the `RosEnv` interface.